sdks/java/io/tika/src/main/java/org/apache/beam/sdk/io/tika/TikaSource.java

```
@@ -0,0 +1,466 @@
+        } finally {
+          is.close();
+        }
+      }
+      return advanceToNext();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      checkState(current != null, "Call start() before advance()");
+      return advanceToNext();
+    }
+
+    protected boolean advanceToNext() throws IOException {
+      current = null;
+      // The content is reported first
+      if (metadataIterator == null) {
+        // Check if some content is already available
+        current = tikaHandler.getCurrent();
+
+        if (current == null && !Boolean.TRUE.equals(spec.getParseSynchronously())) {
+          long maxPollTime = 0;
+          long configuredMaxPollTime = spec.getQueueMaxPollTime() == null
+              ? TikaIO.Read.DEFAULT_QUEUE_MAX_POLL_TIME : spec.getQueueMaxPollTime();
+          long configuredPollTime = spec.getQueuePollTime() == null
+              ? TikaIO.Read.DEFAULT_QUEUE_POLL_TIME : spec.getQueuePollTime();
+
+          // Poll the queue till the next piece of data is available
+          while (current == null && maxPollTime < configuredMaxPollTime) {
+            boolean docEnded = tikaHandler.waitForNext(configuredPollTime);
+            current = tikaHandler.getCurrent();
+            // End of Document ?
+            if (docEnded) {
+              break;
+            }
+            maxPollTime += spec.getQueuePollTime();
+          }
+        }
+        // No more content ?
+        if (current == null && Boolean.TRUE.equals(spec.getReadOutputMetadata())) {
+          // Time to report the metadata
+          metadataIterator = Arrays.asList(tikaMetadata.names()).iterator();
+        }
+      }
+
+      if (metadataIterator != null && metadataIterator.hasNext()) {
+          String key = metadataIterator.next();
+          // The metadata name/value separator can be configured if needed
+          current = key + "=" + tikaMetadata.get(key);
+      }
+      return current != null;
```

It seems like we will arbitrarily truncate documents and silently lose data if polling takes more time than specified - this falls under "Never, ever, ever do this" in https://beam.apache.org/contribute/ptransform-style-guide/#runtime-errors-and-data-consistency

---

TikaReader does its best not to be the blocking point in the pipeline. One can set this value to a very large number and to ensure that any given file will always be parsed, it's not as if we get some unbounded data from the network, etc...And one can choose to parse synchronously